File: Server_Root/Server/RoomServer/Messages/GameEventIdleToR.cpp
#include "stdafx.h"
#include "GameEventIdleToR.h"
#include "GameEventIdleToR_Data_generated.h"

GameEventIdleToR::GameEventIdleToR() : m_Builder(1024)
{
}

GameEventIdleToR::~GameEventIdleToR()
{
}

unsigned short GameEventIdleToR::GetID()
{
	return MESSAGE_ID;
}

IMessage* GameEventIdleToR::Clone()
{
	GameEventIdleToR* pClone = new GameEventIdleToR();

	pClone->m_nPlayerIndex = m_nPlayerIndex;
	pClone->m_vec3Pos = m_vec3Pos;

	return pClone;
}

const char* GameEventIdleToR::Serialize(int* pLength)
{
	FBSData::Vector3 pos(m_vec3Pos.x, m_vec3Pos.y, m_vec3Pos.z);

	GameEventIdleToR_DataBuilder data_builder(m_Builder);
	data_builder.add_PlayerIndex(m_nPlayerIndex);
	data_builder.add_Pos(&pos);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool GameEventIdleToR::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<GameEventIdleToR_Data>((const void*)pChar);

	m_nPlayerIndex = data->PlayerIndex();
	m_vec3Pos.x = data->Pos()->x();
	m_vec3Pos.y = data->Pos()->y();
	m_vec3Pos.z = data->Pos()->z();

	return true;
}

File: Client_Root/Client/Assets/Scripts/Network/Messages/ToRoom/GameEventIdleToR.cs
using UnityEngine;
using FlatBuffers;

public class GameEventIdleToR : IMessage
{
    public const ushort MESSAGE_ID = MessageID.GameEventIdleToR_ID;

    public int m_nPlayerIndex;
    public Vector3 m_vec3Pos;

    public ushort GetID()
    {
        return MESSAGE_ID;
    }

    public IMessage Clone()
    {
        return null; 
    }

    public byte[] Serialize()
    {
        FlatBufferBuilder builder = new FlatBufferBuilder(1024);

        var pos = FBSData.Vector3.CreateVector3(builder, m_vec3Pos.x, m_vec3Pos.y, m_vec3Pos.z);

        GameEventIdleToR_Data.StartGameEventIdleToR_Data(builder);
        GameEventIdleToR_Data.AddPlayerIndex(builder, m_nPlayerIndex);
        GameEventIdleToR_Data.AddPos(builder, pos);
        var data = GameEventIdleToR_Data.EndGameEventIdleToR_Data(builder);

        builder.Finish(data.Value);

        return builder.SizedByteArray();
    }

    public bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = GameEventIdleToR_Data.GetRootAsGameEventIdleToR_Data(buf);

        m_nPlayerIndex = data.PlayerIndex;
        m_vec3Pos = new Vector3(data.Pos.X, data.Pos.Y, data.Pos.Z);

        return true;
    }
}
